bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalForceTest.java

```diff
@@ -0,0 +1,405 @@
+        journal.start();
+
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicInteger countForceWriteCallsBeforeAck = new AtomicInteger();
+        replace(method(JournalChannel.class, "forceWrite")).with(new InvocationHandler() {
+            @Override
+            public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                log.info("{} called with arguments {} ", method.getName(), Arrays.toString(args));
+                if (latch.getCount() == 1) {
+                    countForceWriteCallsBeforeAck.incrementAndGet();
+                }
+                try {
+                    method.invoke(proxy, args);
+                } catch (InvocationTargetException err) {
+                    throw err.getCause();
+                }
+                return null;
+            }
+        });
+
+        // wait for first file to be created
+        while (countForceWriteCallsBeforeAck.get() == 0) {
+            Thread.sleep(100);
+        }
+
+        long ledgerId = 1;
+        long entryId = 0;
+        journal.logAddEntry(ledgerId, entryId, DATA, false /* ackBeforeSync */, new WriteCallback() {
+            @Override
+            public void writeComplete(int rc, long ledgerId, long entryId, BookieSocketAddress addr, Object ctx) {
+                latch.countDown();
+            }
+        }, null);
+        assertTrue(latch.await(20, TimeUnit.SECONDS));
+        assertEquals(2, countForceWriteCallsBeforeAck.get());
+
+        journal.shutdown();
+    }
+
+    @Test
+    public void testAckBeforeSyncSuspendingJournal() throws Exception {
+        File journalDir = tempDir.newFolder();
+        Bookie.checkDirectoryStructure(Bookie.getCurrentDirectory(journalDir));
+
+        ServerConfiguration conf = TestBKConfiguration.newServerConfiguration();
+        conf.setJournalDirName(journalDir.getPath())
+            .setZkServers(null);
+
+        LedgerDirsManager ledgerDirsManager = mock(LedgerDirsManager.class);
+        Journal journal = new Journal(journalDir, conf, ledgerDirsManager);
+
+        CountDownLatch forceWriteThreadSuspendedIsWaiting = new CountDownLatch(1);
+        LinkedBlockingQueue<ForceWriteRequest> supportQueue = new LinkedBlockingQueue<>();
+        @SuppressWarnings("unchecked")
+        BlockingQueue<ForceWriteRequest> forceWriteRequests = mock(BlockingQueue.class);
+        AtomicBoolean forceWriteThreadSuspended = new AtomicBoolean(true);
+
+        doAnswer((Answer) (InvocationOnMock iom) -> {
+            supportQueue.put(iom.getArgument(0));
+            return null;
+        }).when(forceWriteRequests).put(any(ForceWriteRequest.class));
+
+        when(forceWriteRequests.take()).thenAnswer(new Answer<ForceWriteRequest>() {
+            @Override
+            public ForceWriteRequest answer(InvocationOnMock iom) throws Throwable {
+                forceWriteThreadSuspendedIsWaiting.countDown();
```

Comment:
you don't really need `forceWriteThreadSuspendedIsWaiting` here. because what you want to test here is `forceWrite` is never called, when ackBeforeSync is completed. and you don't really need another supportQueue.

```
CountDownLatch forceWriteThreadSuspendedLatch = new CountDownLatch(1);
BlockingQueue<ForceWriteRequest> forceWriteRequests = mock(BlockingQueue.class);
when(forceWriteRequests.take()).thenAnswer(invocationOnMock -> {
    // suspend the force write thread
    forceWriteThreadSuspendedLatch.await();

   return invocationOnMock.callRealMethod();
});

Whitebox.setInternalState(journal, "forceWriteRequests", forceWriteRequests);

// mock journal channel ...

journal.start();

journal.logAddEntry(...); // add an entry with `ackBeforeSync = true`

// wait for addentry to complete
// verify forceWrite is never called.
verify(jc, times(0)).forceWrite(...);

// resume the forcewrite queue
forceWriteThreadSuspendedLatch.countDown();
```

Reply:
@sijie your approach is not working because a mock BlockingQueue is not enoufh, the ForceWriteThread needs a real queue, which retains entries and finally returns them.
And we cannot use callRealMethod() on a "interface"

Reply:
@eolivelli sorry it was my typo. you should be using `spy`.

`BlockingQueue<ForceWriteRequest> forceWriteRequests = spy(new LinkedBlockingQueue())`